### ISSUE TITLE
Empirical overlap != Density Overlap ?

### ISSUE THREAD

**Issue by russellpierce** [masspastore/overlapping#10]
It might be that I am misunderstanding your package.  My intuition was that a uniform distribution [0,1] should overlap with a uniform distribution [0.5, 1.5] by 50%.  However...
```
result <- replicate(1000, {
  x <- list(X1 = runif(100000), X2 = runif(10000, min = .5, max=1.5))
  out <- overlap( x, plot = FALSE)
  out$OV # estimated overlapped areas
})
mean(result)
median(result)
```
.... shows a bias towards indicating a high proportion of overlap than 50%.  Looking at the graphical output and skimming the code I gather that this is because you're slicing up the results of a call to `density` to find overlap.  However, `density` applies a gaussian kernel that can't adequately represent the sharp edges of the uniform distributions.  After applying a density function with a kernel like that to what extent can we still call the overlap 'empirical'?

**Comment by masspastore** [comment]
Thanks @russellpierce, I agree with you.

The problem depends on the `density` function, you can verify this with the following code. You will observe that `x` values returned are outside natural limits 0-1

`set.seed(1)`
`density(runif(1000))`

And consequently this had an influence on the overlapping estimation. 

In order to fix this problem, I changed the function `overlap` by allowing to add all additional arguments of `density` function (for example the choice of kernel) and, in particular,  the possibility of defining the empirical bound of input variables.  I think that the estimate is much more improved now, even though it's still an approximation. In my experience (working with psychological and social data), this shouldn't be particularly problematic.

**Comment by russellpierce** [comment]
I agree that the problem comes from the density function and I commend your improvements.  

I acknowledge that you've almost certainly thought about this more than me.  So, what I'm trying to do here is get to a place where I can feel 👍 about saying 'works as expected' and/or we both learn something.  Thank you for your patience.

I'm still a bit stuck about whether this procedure reflects empirical overlap when data has to go through a density function of any kind.  From your description, what I'd expect would be a bit more like...

```
library(dplyr)
d <- list(X1 = runif(1000000), X2 = runif(100000, min = .5, max=1.5))
# as with your function, we need a set of bins to evaluate the result under
nbins <- 1000

# These are function generators, making an ECDF for each dataset
emperical_1 <- ecdf(d$X1)
emperical_2 <- ecdf(d$X2)

# Establishing the range of values to build bins over
min_val <- min(c(d$X1, d$X2))
max_val <- max(c(d$X1, d$X2))

# ecdf yields a cumulative mass
cumulative_mass_1 <- lapply(seq(min_val, max_val, length.out = nbins), emperical_1)
cumulative_mass_2 <- lapply(seq(min_val, max_val, length.out = nbins), emperical_2)

# The mass in any given bin is the difference between the cumulative at the end of this bin and the cumulative at the end of the previous bin.
bin_mass_1 <- unlist(cumulative_mass_1)-dplyr::lag(unlist(cumulative_mass_1))
bin_mass_2 <- unlist(cumulative_mass_2)-dplyr::lag(unlist(cumulative_mass_2))

# The overlap is the sum of the lower mass all bin pairs
sum(pmin(bin_mass_1, bin_mass_2), na.rm = TRUE)
```

... but admittedly that skews low in overlap estimation.  Can you describe a bit about why the function you propose is better in practice / more intuitive over something like what I've sketched above?

**Comment by masspastore** [comment]
thanks @russellpierce, 

I updated everything following your suggestions. Now I hope it's fine.

**Comment by soodoku** [comment]
hey @masspastore: I have signed off for JOSS. But I think it may be useful to highlight that the results being produced are approximate in the readme. And to also note any systematic issues. I would recommend just using the examples here to illustrate the issues.  Problem is = a bunch of people will just use the package without being aware of the issues if you don't put it in a prominent place.

**Comment by masspastore** [comment]
@soodoku:  I updated the readme file by adding a note about this issue.  Thanks.
